### DISCUSSION TITLE
Sonoff S31 leds not properly indicating status

### DISCUSSION THREAD

**Discussion by shir474** [arendst/Tasmota#16128] [Q&A] [answered]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._

Program Version: 12.0.2(tasmota)
Build Date & Time: 2022-06-20T12:41:03
Core/SD Version: 2_7_4_9/2.2.2-dev(38a443e)

Sonoff S31 has two led lights (red and blue). If I use module type Sonoff S31 (41) and Template Sonoff S31 (41), no matter how I configure GPIO13 LED/LED_i, the two leds come on when the relay is on and both go off when the relay is off. However, if I use Generic (0) module type and Sonoff S31 (41) template and configure GPIO13 LED 2, the blue led goes on when the device boots. When the relay is on blue led goes off and red led goes on. When I turn the relay off blue led goes on and red led goes off. This is closer to the expected behavior. There must be something different with the Sonoff S31 Module type not supporting the GPIO13 LEDs properly.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [X] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [X] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [X] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [X] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [X] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [X] Device used (e.g., Sonoff Basic): _____ Sonoff S31
- [X] Tasmota binary firmware version number used: _____ 12.0.2(tasmota)
  - [X] Pre-compiled
  - [ ] Self-compiled
- [X] Flashing tools used: _____ esptool
- [X] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:

18:16:02.915 MQT: stat/XXXXX/RESULT = {"NAME":"Generic","GPIO":[32,3072,0,3104,0,0,0,0,224,289,0,0,0,0],"FLAG":0,"BASE":41}
18:16:03.114 MQT: stat/XXXXX/RESULT = {"Module":{"0":"Generic"}}
18:16:03.320 MQT: stat/XXXXX/RESULT = {"GPIO0":{"32":"Button1"},"GPIO1":{"3072":"CSE7766 Tx"},"GPIO2":{"0":"None"},"GPIO3":{"3104":"CSE7766 Rx"},"GPIO4":{"0":"None"},"GPIO5":{"0":"None"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"224":"Relay1"},"GPIO13":{"289":"Led2"},"GPIO14":{"0":"None"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}

```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [X] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:

18:17:48.560 MQT: stat/XXXXX/STATUS = {"Status":{"Module":0,"DeviceName":"XXXXX","FriendlyName":["XXXXX"],"Topic":"XXXXX","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":0,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
18:17:48.566 MQT: stat/XXXXX/STATUS1 = {"StatusPRM":{"Baudrate":4800,"SerialConfig":"8E1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Software/System restart","Uptime":"0T18:17:50","StartupUTC":"","Sleep":50,"CfgHolder":4617,"BootCount":71,"BCResetTime":"1970-01-01T00:00:00","SaveCount":84,"SaveAddress":"F9000"}}
18:17:48.570 MQT: stat/XXXXX/STATUS2 = {"StatusFWR":{"Version":"12.0.2(tasmota)","BuildDateTime":"2022-06-20T12:41:03","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"329/699"}}
18:17:48.574 MQT: stat/XXXXX/STATUS3 = {"StatusLOG":{"SerialLog":0,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["XXXXX",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000680003C5A0A190000000000","00000080","00006000","00004000","00000000"]}}
18:17:48.585 MQT: stat/XXXXX/STATUS4 = {"StatusMEM":{"ProgramSize":619,"Free":384,"Heap":22,"ProgramFlashSize":1024,"FlashSize":4096,"FlashChipId":"16405E","FlashFrequency":40,"FlashMode":3,"Features":["00000809","8F9AC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","14000020","00000000"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45,62","Sensors":"1,2,3,4,5,6"}}
18:17:48.593 MQT: stat/XXXXX/STATUS5 = {"StatusNET":{"Hostname":"XXXXX","IPAddress":"XXXXX","Gateway":"XXXXX","Subnetmask":"255.255.255.0","DNSServer1":"1XXXXX","DNSServer2":"0.0.0.0","Mac":"XXXXX","Webserver":2,"HTTP_API":1,"WifiConfig":2,"WifiPower":17.0}}
18:17:48.597 MQT: stat/XXXXX/STATUS6 = {"StatusMQT":{"MqttHost":"XXXXX","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"XXXXX","MqttUser":"XXXXX","MqttCount":18,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
18:17:48.604 MQT: stat/XXXXX/STATUS7 = {"StatusTIM":{"UTC":"1970-01-01T18:17:48","Local":"1970-01-01T18:17:48","StartDST":"1970-01-01T00:00:00","EndDST":"1970-01-01T00:00:00","Timezone":"+00:00","Sunrise":"20:13","Sunset":"05:47"}}
18:17:48.607 MQT: stat/XXXXX/STATUS9 = {"StatusPTH":{"PowerDelta":[0,0,0],"PowerLow":0,"PowerHigh":0,"VoltageLow":0,"VoltageHigh":0,"CurrentLow":0,"CurrentHigh":0}}
18:17:48.615 MQT: stat/XXXXX/STATUS10 = {"StatusSNS":{"Time":"1970-01-01T18:17:48","ENERGY":{"TotalStartTime":"1970-01-01T00:00:00","Total":0.113,"Yesterday":0.000,"Today":0.113,"Power": 0,"ApparentPower": 0,"ReactivePower": 0,"Factor":0.00,"Voltage": 0,"Current":0.000}}}
18:17:48.621 MQT: stat/XXXXX/STATUS11 = {"StatusSTS":{"Time":"1970-01-01T18:17:48","Uptime":"0T18:17:50","UptimeSec":65870,"Heap":22,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":18,"POWER":"OFF","Wifi":{"AP":1,"SSId":"XXXXX","BSSId":"XXXXXX","Channel":11,"Mode":"11n","RSSI":100,"Signal":-39,"LinkCount":1,"Downtime":"0T00:00:03"}}}

```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_

See problem desription

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

Blue led be status (wifi) and red led to be relay on/off indicator

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by wayner9** [comment]
I have the same question, or a very similar question.  I just flashed some S31s with Tasmota and used the S31 template.  The two LEDs seem to be tied together.  When the relay is on they are both on, when the relay is off they are both off.  But I believe the "proper" way for them to operate is for the blue LED to always be on as it is a power indicator and the red LED should show the relay state - off when the relay is off and on/red when the relay is on.

How can I change this?  I guess what I want to change is the behaviour of the blue LED as the red LED is behaving in the expected manner.

**Reply by barbudor** [reply]
From the template [here](https://templates.blakadder.com/sonoff_S31.html) there is only 1 led under control of Tasmota. The other may be directly driven by the same gpio as the relay, so it will always shows the relay state. 

An option could be to set GPIO13 as LedLink_i (instead of Led_i) to display the connectios status. See the doc about [LedLink](https://tasmota.github.io/docs/Lights/#status-leds)

**Reply by wayner9** [reply]
Thanks for the response.  If I wanted to change GPIO13 to LedLink_i then can I just go to the Template Parameters page and change the option for GPIO13 from Led_i to LedLink_i?  Or does that just create a new template and not actually change the settings on the device?
edit - I tried that and it just keeps the blue LED permanently off.  Unless I need to configure something else.

**Reply by sfromis** [reply]
Tasmota only has one template, and you can play around to see what works for you. A pin configured as LedLink (or _i) will typically blink right after boot until network connections established, and hopefully stay off. You can invert it if you prefer it the other way around.

**Reply by wayner9** [reply]
Thanks - How do I invert it?  I like the idea of a blue LED meaning that it is connected to Wifi.  Currently I am not using MQTT, just Wifi commands so the connectivity that I care about is only Wifi.  I saw that some of the status indicators talk about MQTT and Wifi.

**Reply by sfromis** [reply, marked as answer]
You already know how to invert :wink: It is the `_i` which you can set or remove. And when I deliberately said "network" instead of "Wifi" it is because it includes MQTT, but only if that feature is enabled.

**Reply by wayner9** [reply]
Perfect - changing GPIO13 to LedLink is the setting that I wanted.  Thanks for your help!
